### PR TITLE
Ability to constrain VPA MaxAllowed per seed

### PR DESCRIPTION
**Pull request by andrerun**
**How to categorize this PR?**
/area auto-scaling
/kind enhancement

**What this PR does / why we need it**:
Adds a seed setting which caps VPA MaxAllowed for shoot control plane components.

The newly added, optional seed setting `Seed.Spec.Settings.VerticalPodAutoscaler.MaxAllowed` controls the `MaxAllowed` value for shoot control plane VPAs. In the presence of that setting:
- If a component has a default VPA MaxAllowed for a given resource (cpu/memory), and the new seed-level setting also prescribes a value for that resource, gardenlet will set the smaller of the two values in the component's actual VPA object.
- If only one of the two sources prescribes a MaxAllowed value, that value will be used.
The new seed-level setting currently affects `etcd`, `kube-apiserver`, and `prometheus`. The vision is to follow up with support for the rest of the shoot control plane components, and for garden control plane components.

**Special notes for your reviewer**:

**Release note**:

```feature operator
`spec.settings.verticalPodAutoscaler.maxAllowed` setting was added to the `Seed` API. It allows constraining the Vertical Pod Autoscaler `maxAllowed` value for shoot control plane components (currently affects `etcd`, `kube-apiserver`, and `prometheus`).
```

### CODE REVIEW

File: example/50-seed.yaml [hunk 1]
```diff
@@ -81,6 +81,10 @@ spec:
   #       allowed: true
     verticalPodAutoscaler:
       enabled: true # a Gardener-managed VPA deployment is enabled
+#     maxAllowed: # Sets shoot components' VPA MaxAllowed. For each resource (cpu/memory) the resulting value will be the
```
Comment (rfranzke) [inline example/50-seed.yaml:84]: Please augment https://github.com/gardener/gardener/blob/master/docs/operations/seed_settings.md#vertical-pod-autoscaler accordingly

File: example/50-seed.yaml [hunk 2]
```diff
@@ -81,6 +81,10 @@ spec:
...
+#                 # lesser of the default MaxAllowed (if such default exists), and the value specified here.
+#        cpu: 8
+#        memory: 32Gi
```
Comment (rfranzke) [inline example/50-seed.yaml:87]:
```suggestion
#       cpu: 8
#       memory: 32Gi
```

File: pkg/apis/core/types_seed.go [hunk 3]
```diff
@@ -274,6 +274,15 @@ type SeedSettingVerticalPodAutoscaler struct {
...
 	Enabled bool
+
+	// MaxAllowed specifies the maximum amount of resources that will be recommended for Gardener components. The value is applied
+	// by setting it in the MaxAllowed field of the component's VPA object. If the VPA has a default value
+	// for MaxAllowed, the smaller of the two values is used. If this field is nil, no change is made to the MaxAllowed
+	// values in component VPAs.
+	//
+	// The effects of the setting are currently limited to the following shoot control plane components:
+	//  etcd, kube-apiserver,prometheus
+	MaxAllowed corev1.ResourceList
```
Comment (rfranzke) [inline pkg/apis/core/types_seed.go:285]:
```suggestion
	MaxAllowed *corev1.ResourceList
```

File: pkg/apis/core/v1beta1/types_seed.go [hunk 4]
```diff
@@ -341,6 +341,16 @@ type SeedSettingVerticalPodAutoscaler struct {
...
+
+	// MaxAllowed specifies the maximum amount of resources that will be recommended for Gardener components. The value is applied
+	// by setting it in the MaxAllowed field of the component's VPA object. If the VPA has a default value
+	// for MaxAllowed, the smaller of the two values is used. If this field is nil, no change is made to the MaxAllowed
+	// values in component VPAs.
+	//
+	// The effects of the setting are currently limited to the following shoot control plane components:
+	//  etcd, kube-apiserver,prometheus
+	// +optional
+	MaxAllowed corev1.ResourceList `json:"maxAllowed,omitempty" protobuf:"bytes,2,rep,name=maxAllowed,casttype=k8s.io/api/core/v1.ResourceList,castkey=k8s.io/api/core/v1.ResourceName"`
```
Comment (rfranzke) [inline pkg/apis/core/v1beta1/types_seed.go:353]:
```suggestion
	MaxAllowed *corev1.ResourceList `json:"maxAllowed,omitempty" protobuf:"bytes,2,rep,name=maxAllowed,casttype=k8s.io/api/core/v1.ResourceList,castkey=k8s.io/api/core/v1.ResourceName"`
```

File: pkg/component/etcd/etcd/etcd_test.go [hunk 5]
```diff
@@ -1369,6 +1404,122 @@ var _ = Describe("Etcd", func() {
 			Expect(etcd.Deploy(ctx)).To(Succeed())
 		})
 
+		Context("When VPA MaxAllowed seed level setting is in effect", func() {
+			BeforeEach(func() {
+				vpaEnabled = true
+			})
+			It("should successfully deploy (normal etcd) and use VPA MaxAllowed from the seed level setting", func() {
```
Comment (rfranzke) [inline pkg/component/etcd/etcd/etcd_test.go:1411]:
```suggestion

			It("should successfully deploy (normal etcd) and use VPA MaxAllowed from the seed level setting", func() {
```

File: pkg/gardenlet/operation/botanist/etcd.go [hunk 6]
```diff
@@ -47,6 +47,15 @@ func (b *Botanist) DefaultEtcd(role string, class etcd.Class) (etcd.Interface, e
...
 	}
 
+	var vpaMaxAllowed corev1.ResourceList
+	if b.Config != nil &&
+		b.Config.SeedConfig != nil &&
+		b.Config.SeedConfig.Spec.Settings != nil &&
+		b.Config.SeedConfig.Spec.Settings.VerticalPodAutoscaler != nil &&
+		b.Config.SeedConfig.Spec.Settings.VerticalPodAutoscaler.MaxAllowed != nil {
+		vpaMaxAllowed = b.Config.SeedConfig.Spec.Settings.VerticalPodAutoscaler.MaxAllowed.DeepCopy()
+	}
```
Comment (rfranzke) [inline pkg/gardenlet/operation/botanist/etcd.go:57]: I would rather expect `MergeMinValuesIntoResourceList` to make a copy and not mutate the passed argument. Callers shouldn't care.

File: pkg/gardenlet/operation/botanist/kubeapiserver.go [hunk 7]
```diff
@@ -136,6 +136,15 @@ func (b *Botanist) computeKubeAPIServerAutoscalingConfig() apiserver.Autoscaling
...
 	}
 
+	var vpaMaxAllowed corev1.ResourceList
+	if b.Config != nil &&
+		b.Config.SeedConfig != nil &&
+		b.Config.SeedConfig.Spec.Settings != nil &&
+		b.Config.SeedConfig.Spec.Settings.VerticalPodAutoscaler != nil &&
+		b.Config.SeedConfig.Spec.Settings.VerticalPodAutoscaler.MaxAllowed != nil {
+		vpaMaxAllowed = b.Config.SeedConfig.Spec.Settings.VerticalPodAutoscaler.MaxAllowed.DeepCopy()
+	}
```
Comment (rfranzke) [inline pkg/gardenlet/operation/botanist/kubeapiserver.go:146]: same here

File: pkg/gardenlet/operation/botanist/monitoring.go [hunk 8]
```diff
@@ -82,6 +83,15 @@ func (b *Botanist) DefaultPrometheus() (prometheus.Interface, error) {
...
 	}
 
+	var vpaMaxAllowed corev1.ResourceList
+	if b.Config != nil &&
+		b.Config.SeedConfig != nil &&
+		b.Config.SeedConfig.Spec.Settings != nil &&
+		b.Config.SeedConfig.Spec.Settings.VerticalPodAutoscaler != nil &&
+		b.Config.SeedConfig.Spec.Settings.VerticalPodAutoscaler.MaxAllowed != nil {
+		vpaMaxAllowed = b.Config.SeedConfig.Spec.Settings.VerticalPodAutoscaler.MaxAllowed.DeepCopy()
+	}
```
Comment (rfranzke) [inline pkg/gardenlet/operation/botanist/monitoring.go:93]: same here

File: pkg/operator/controller/garden/garden/components.go [hunk 9]
```diff
@@ -518,6 +518,7 @@ func (r *Reconciler) newEtcd(
 			HighAvailabilityEnabled:     highAvailabilityEnabled,
 			TopologyAwareRoutingEnabled: helper.TopologyAwareRoutingEnabled(garden.Spec.RuntimeCluster.Settings),
 			VPAEnabled:                  features.DefaultFeatureGate.Enabled(features.VPAForETCD),
+			VPAMaxAllowed:               nil,
```
Comment (rfranzke) [inline pkg/operator/controller/garden/garden/components.go:521]: This line should not be needed, or?

```suggestion
```

File: pkg/utils/kubernetes/resourcelist_test.go [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+import (
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+
+	. "github.com/gardener/gardener/pkg/utils/kubernetes"
+)
+
+var _ = Describe("Node", func() {
```
Comment (rfranzke) [inline pkg/utils/kubernetes/resourcelist_test.go:16]:
```suggestion
var _ = Describe("ResourceList", func() {
```

File: pkg/utils/kubernetes/resourcelist_test.go [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+	Describe("IsNodeLabelAllowedForKubelet", func() {
```
Comment (rfranzke) [inline pkg/utils/kubernetes/resourcelist_test.go:17]:
```suggestion
	Describe("#MergeMinValuesIntoResourceList", func() {
```

**Review by rfranzke — CHANGES_REQUESTED** [review]
Don't we need a similar settings here? https://github.com/gardener/gardener/blob/8d81563f60891bdfb0947b05e6216ffd1635cfe3/example/operator/20-garden.yaml#L39-L40

File: pkg/apis/core/types_seed.go (hunk 3 above)
Comment (ialidzhikov) [reply pkg/apis/core/types_seed.go:285]: corev1.ResourceList is a type alias to `map[ResourceName]resource.Quantity`. Ref https://github.com/kubernetes/kubernetes/blob/0a406ee1b4b65574c02f8aec6ff5a69d07b68be6/staging/src/k8s.io/api/core/v1/types.go#L6340

Hence, we don't need to use a pointer. It is already a reference type.

**Comment by ialidzhikov** [comment]
I will take over the changes and raise a new PR.

/close

**Pull request closed without merging**